src/v/utils/stream_utils.h

```diff
@@ -161,19 +165,24 @@ class input_stream_fanout final {
     /// the resulting buffers to clients
     ss::future<> produce() {
         gate_guard g(_gate);
-        while (!_in.eof() && !_gate.is_closed()) {
-            auto units = co_await ss::get_units(_sem, 1);
-            ss::temporary_buffer<Ch> buf;
-            if (_max_size == 0) {
-                buf = co_await _in.read();
-            } else {
-                buf = co_await _in.read_up_to(_max_size);
+        try {
+            while (!_in.eof() && !_gate.is_closed()) {
+                auto units = co_await ss::get_units(_sem, 1);
+                ss::temporary_buffer<Ch> buf;
+                if (_max_size == 0) {
+                    buf = co_await _in.read();
+                } else {
+                    buf = co_await _in.read_up_to(_max_size);
+                }
+                ++_cnt;
+                _buffer.push_back(data_item{
+                  .buf = std::move(buf),
+                  .units = std::move(units),
+                  .mask = _bitmask});
+                _pcond.broadcast();
             }
-            ++_cnt;
-            _buffer.push_back(data_item{
-              .buf = std::move(buf),
-              .units = std::move(units),
-              .mask = _bitmask});
+        } catch (...) {
+            _producer_error = std::current_exception();
             _pcond.broadcast();
```

Comment:
This needs to be `_pcond.broken()` since the waiter is also waiting for a predicate that won't complete.